<a href="https://colab.research.google.com/github/phamvanhoangphi13112002/SuperstoreSalesAnalysis/blob/main/EDA%2BCleaning_Superstore_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Import necessary libraries**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
import os

### **Read file to Dataframes**

In [2]:
file_path = r'D:\JOBBBB\PORTFOLIO\Sale analysis\1. Data source\Global-Superstore.xlsx'

# Read each sheet
df1 = pd.read_excel(file_path, sheet_name='Orders')
df2 = pd.read_excel(file_path, sheet_name='Returns')
df3 = pd.read_excel(file_path, sheet_name='People')

## **Sheet Orders**

### **Overview**

In [3]:
pd.set_option('display.max_columns', None)
df1.head()

Row ID                Order ID Order Date  Ship Date       Ship Mode  \
0   24599  IN-2017-CA120551-42816 2017-03-22 2017-03-29  Standard Class   
1   29465  ID-2015-BD116051-42248 2015-09-01 2015-09-04    Second Class   
2   24598  IN-2017-CA120551-42816 2017-03-22 2017-03-29  Standard Class   
3   24597  IN-2017-CA120551-42816 2017-03-22 2017-03-29  Standard Class   
4   29464  ID-2015-BD116051-42248 2015-09-01 2015-09-04    Second Class   

  Customer ID    Customer Name      Segment  Postal Code   City  State  \
0   CA-120551  Cathy Armstrong  Home Office          NaN  Herat  Hirat   
1   BD-116051     Brian Dahlen     Consumer          NaN  Herat  Hirat   
2   CA-120551  Cathy Armstrong  Home Office          NaN  Herat  Hirat   
3   CA-120551  Cathy Armstrong  Home Office          NaN  Herat  Hirat   
4   BD-116051     Brian Dahlen     Consumer          NaN  Herat  Hirat   

       Country         Region        Market   Product ID  \
0  Afghanistan  Southern Asia  Asia Pacific  FUR-BO-4861   
1  Afghanistan  Southern Asia  Asia Pacific  OFF-SU-2988   
2  Afghanistan  Southern Asia  Asia Pacific  TEC-MA-4211   
3  Afghanistan  Southern Asia  Asia Pacific  FUR-FU-5726   
4  Afghanistan  Southern Asia  Asia Pacific  OFF-EN-3664   

                                        Product Name Sub-Category  \
0                    Ikea Library with Doors, Mobile    Bookcases   
1                           Acme Scissors, Easy Grip     Supplies   
2                       Epson Receipt Printer, White     Machines   
3                    Rubbermaid Door Stop, Erganomic  Furnishings   
4  Cameo Interoffice Envelope, with clear poly wi...    Envelopes   

          Category   Sales  Quantity  Discount  Profit  Shipping Cost  \
0        Furniture  731.82         2       0.0  102.42          39.66   
1  Office Supplies  243.54         9       0.0  104.49          18.72   
2       Technology  346.32         3       0.0   13.77          14.10   
3        Furniture  169.68         4       0.0   79.68          11.01   
4  Office Supplies  203.88         4       0.0   24.36           5.72   

  Order Priority  
0         Medium  
1         Medium  
2         Medium  
3         Medium  
4         Medium

In [4]:
print("The Orders table has:",df1.shape[0]," rows và ", df1.shape[1]," columns")

The Orders table has: 51290  rows và  24  columns


In [5]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51290 entries, 0 to 51289
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          51290 non-null  int64         
 1   Order ID        51290 non-null  object        
 2   Order Date      51290 non-null  datetime64[ns]
 3   Ship Date       51290 non-null  datetime64[ns]
 4   Ship Mode       51290 non-null  object        
 5   Customer ID     51290 non-null  object        
 6   Customer Name   51290 non-null  object        
 7   Segment         51290 non-null  object        
 8   Postal Code     9994 non-null   float64       
 9   City            51290 non-null  object        
 10  State           51290 non-null  object        
 11  Country         51290 non-null  object        
 12  Region          51290 non-null  object        
 13  Market          51290 non-null  object        
 14  Product ID      51290 non-null  object        
 15  Pr

### **Check null value**

In [6]:
df1.isnull().values.any()

True

In [7]:
df1.isnull().sum()

Row ID                0
Order ID              0
Order Date            0
Ship Date             0
Ship Mode             0
Customer ID           0
Customer Name         0
Segment               0
Postal Code       41296
City                  0
State                 0
Country               0
Region                0
Market                0
Product ID            0
Product Name          0
Sub-Category          0
Category              0
Sales                 0
Quantity              0
Discount              0
Profit                0
Shipping Cost         0
Order Priority        0
dtype: int64

* The field "Postal Code" has 41296 null values, so it will be drop out of this data

### **Check duplicated values**

In [8]:
df1.duplicated().sum()

0

* There is no duplicated values in this dataset


### **Check unique value for each field**

In [9]:
for i in df1.columns:
  print('Number of unique value in', i,'is:',df1[i].nunique())

Number of unique value in Row ID is: 51290
Number of unique value in Order ID is: 25728
Number of unique value in Order Date is: 1429
Number of unique value in Ship Date is: 1463
Number of unique value in Ship Mode is: 4
Number of unique value in Customer ID is: 17415
Number of unique value in Customer Name is: 796
Number of unique value in Segment is: 3
Number of unique value in Postal Code is: 631
Number of unique value in City is: 3650
Number of unique value in State is: 1106
Number of unique value in Country is: 165
Number of unique value in Region is: 23
Number of unique value in Market is: 5
Number of unique value in Product ID is: 3788
Number of unique value in Product Name is: 3788
Number of unique value in Sub-Category is: 17
Number of unique value in Category is: 3
Number of unique value in Sales is: 22995
Number of unique value in Quantity is: 14
Number of unique value in Discount is: 27
Number of unique value in Profit is: 24575
Number of unique value in Shipping Cost is: 1

### **Cleaning Data and Read Data Frame to New files**

The cleaning scripts:
- Delete the Postal Code columns
- Rename all columns to lowercase, and replace the space by underscore if exist
- Round number into 2 decimal
- Join sheet 2 "Return" to get the data about return goods
- Label for each Order ID: Yes by 1 and fill No by 0
- Sort the data by "Row_ID"
- Drop duplicated values


In [10]:
# Join sheet 2 to get the data about return goods (left-join)
df = pd.merge(df1,df2, on='Order ID', how='left')
# Fill the Nan in Returned by No
df['Returned'] = df['Returned'].fillna('No')
# Label
Returned = {
    'Yes': '1',
    'No': '0',
}
df['Returned'] = df['Returned'].replace(Returned)
# Delete PostalCode columns
df =df.drop('Postal Code', axis =1)
df = df.drop('Region_y', axis =1)
# Rename all columns to lowercase
df.rename(columns=lambda x: x.lower().replace(" ", "_"), inplace=True)
# Change the datatype of (OrderDate and ShipDate) to datetime
df['order_date'] = pd.to_datetime(df['order_date'])
df['ship_date'] = pd.to_datetime(df['ship_date'])
# Round number into 2 decimal
df['profit'] = df['profit'].round(2)
df['shipping_cost'] = df['shipping_cost'].round(2)
# Sort the data by "Row_ID
df = df.sort_values(by='row_id')
# Drop duplicated value
df = df.drop_duplicates(keep='first')
# reset row_id
df['row_id'] = range(1, len(df) + 1)
# Save to Data source
# df.to_csv('/content/gdrive/MyDrive/PROJECTS/Project: SuperstoreSales/global_superstore_clean.csv', index=False)

In [11]:
# Save to Data source
save_path = r'D:\JOBBBB\PORTFOLIO\Sale analysis\global_superstore_clean.csv'
directory_path = os.path.dirname(save_path)
# Create the directory if it doesn't exist
os.makedirs(directory_path, exist_ok=True)

# Save the DataFrame to a CSV file in the specified directory
df.to_csv(save_path, index=False)

In [12]:
df

row_id                 order_id order_date  ship_date       ship_mode  \
26336       1  MX-2017-SC2057582-43010 2017-10-02 2017-10-06  Standard Class   
9122        2  MX-2015-KW1657028-42292 2015-10-15 2015-10-20  Standard Class   
9123        3  MX-2015-KW1657028-42292 2015-10-15 2015-10-20  Standard Class   
9129        4  MX-2015-KW1657028-42292 2015-10-15 2015-10-20  Standard Class   
9128        5  MX-2015-KW1657028-42292 2015-10-15 2015-10-20  Standard Class   
...       ...                      ...        ...        ...             ...   
19222   51286   HU-2015-NF838557-42252 2015-09-05 2015-09-07     First Class   
19204   51287   HU-2015-NF838557-42252 2015-09-05 2015-09-07     First Class   
19207   51288   HU-2015-NF838557-42252 2015-09-05 2015-09-07     First Class   
19196   51289   HU-2015-NF838557-42252 2015-09-05 2015-09-07     First Class   
23089   51290  IZ-2014-SZ1003561-41875 2014-08-24 2014-08-28  Standard Class   

      customer_id     customer_name      segment           city  \
26336  SC-2057582      Sonia Cooley     Consumer    Mexico City   
9122   KW-1657028    Kelly Williams     Consumer  Dos Quebradas   
9123   KW-1657028    Kelly Williams     Consumer  Dos Quebradas   
9129   KW-1657028    Kelly Williams     Consumer  Dos Quebradas   
9128   KW-1657028    Kelly Williams     Consumer  Dos Quebradas   
...           ...               ...          ...            ...   
19222   NF-838557  Natalie Fritzler     Consumer       Budapest   
19204   NF-838557  Natalie Fritzler     Consumer       Budapest   
19207   NF-838557  Natalie Fritzler     Consumer       Budapest   
19196   NF-838557  Natalie Fritzler     Consumer       Budapest   
23089  SZ-1003561        Sam Zeldin  Home Office        Baghdad   

                  state   country         region_x        market   product_id  \
26336  Distrito Federal    Mexico  Central America         LATAM  OFF-LA-4658   
9122          Risaralda  Colombia    South America         LATAM  FUR-FU-6238   
9123          Risaralda  Colombia    South America         LATAM  FUR-BO-4845   
9129          Risaralda  Colombia    South America         LATAM  OFF-BI-3720   
9128          Risaralda  Colombia    South America         LATAM  OFF-AR-5905   
...                 ...       ...              ...           ...          ...   
19222          Budapest   Hungary   Eastern Europe        Europe  OFF-BI-3248   
19204          Budapest   Hungary   Eastern Europe        Europe  TEC-AC-5102   
19207          Budapest   Hungary   Eastern Europe        Europe  OFF-AR-3533   
19196          Budapest   Hungary   Eastern Europe        Europe  OFF-PA-4147   
23089           Baghdad      Iraq     Western Asia  Asia Pacific  OFF-BI-2882   

                                    product_name sub-category  \
26336         Hon File Folder Labels, Adjustable       Labels   
9122                        Tenex Clock, Durable  Furnishings   
9123                Ikea 3-Shelf Cabinet, Mobile    Bookcases   
9129                      Cardinal Binder, Clear      Binders   
9128                 Sanford Canvas, Water Color          Art   
...                                          ...          ...   
19222               Avery Binder Covers, Economy      Binders   
19204                  Logitech Flash Drive, USB  Accessories   
19207           Boston Highlighters, Fluorescent          Art   
19196  Enermax Computer Printout Paper, 8.5 x 11        Paper   
23089                 Acco 3-Hole Punch, Durable      Binders   

              category   sales  quantity  discount  profit  shipping_cost  \
26336  Office Supplies   13.08         3       0.0    4.56           2.03   
9122         Furniture  252.16         8       0.0   90.72          13.45   
9123         Furniture  193.28         2       0.0   54.08           9.63   
9129   Office Supplies   35.44         4       0.0    4.96           1.37   
9128   Office Supplies   71.60         2       0.0   11.44           3.79   
...                ...   

In [13]:
df.shape

(51290, 24)